cql3/restrictions/statement_restrictions.cc

```diff
@@ -702,13 +707,13 @@ bool single_column_restriction::slice::is_satisfied_by(const schema& schema,
         return false;
     }
     return cell_value->with_linearized([&] (bytes_view cell_value_bv) {
-        return to_range(_slice, options, _column_def.name_as_text()).contains(
+        return contains_without_wraparound(to_range(_slice, options, _column_def.name_as_text()),
```

Comment:
This is making it in some ways worse. We may decide to suffer it until we merge the rewrite, but let's at least clarify things:

 - a slice (`x >= ? AND x <= ?`) is inherently a (nonwrapping) interval
 - here, you are constructing a wrapping interval, and checking that it is nonwrapping

A clean fix would have been to support empty intervals, and add a constructor that constructs intervals with the lower bound higher than the upper bound as empty intervals.

I understand this happens only with ALLOW FILTERING. Do we know why?

Reply:
The patch itself looks good (as far as circumstances allow)

Reply:
> I understand this happens only with ALLOW FILTERING. Do we know why?

Because of [this](https://sourcegraph.com/github.com/scylladb/scylla@bbd3ed9d47d70207dc122bdd61cd69aaf3ec1a8a/-/blob/cql3/restrictions/single_column_primary_key_restrictions.hh?subtree=true#L438).  Apparently, wrap-around bounds are filtered out by bounds_ranges.

Reply:
I see, thanks.

Reply:
> We may decide to suffer it until we merge the rewrite

What about backports?

Reply:
The suffering is only for those who are pained by ugly code. Users of the ugly code, or its backports, don't notice.

Reply:
Good point (also made by @avikivity [here](https://github.com/scylladb/scylla/pull/5763#discussion_r438646742)).  We may need to add a member like `is_empty` or some such.

Reply:
We need to rewrite interval as 

```c++
class interval {
       bool empty : 1;  // implies !left_bounded and !right_bounded
       bool singular : 1; // implies left_bounded and right_bounded
       bool left_bounded :  1;
       bool left_bound_inclusive : 1;
       bool right_bounded :  1;
       bool right_bound_inclusive : 1;
       maybe<T> left;
       maybe<T> right;  // unused if singular
};
```

However, it's a lot of work to prepare call sites for empty ranges.

Reply:
Actually, I think that the current code can be amended to allow empty ranges without a significant rewrite, if we take advantage of two facts:
1. In some internal functions (like `contains`) we need to be able to check if the interval is empty or not, but in all of these functions we also have access to the comparator, so it can actually be `empty(cmp)`, implemented as `return _interval.is_wrap_around(cmp)`.
2. In some internal functions we need to be able to create empty intervals, but in all of these functions we also have access to arbitrary values of `T` - e.g. in `split_point`, we can use `T split point` parameter to create an empty interval by creating a double-exclusive interval `{bound(split_point, false), bound(split_point, false)}` - with both bounds being non-inclusive, it's a de-facto empty interval, because its wrapping counterpart will be wrap-around.

Then, the advantage is that simply calling `auto interval = wrapping_interval(7, 5)` is a legal operation and creates an empty interval. If we were to use a `_empty` field, then such a constructor would also need to take another parameter - a comparator, which checks if end > start, in which case we set `empty = true`. That's not only ugly, but also close-to-impossible in some use-cases, e.g. when IDL wants to create an interval from a network buffer and it doesn't know or care what its comparator is. Of course, we can also rewrite the IDL layer to send all these boolean values as well as bounds when serializing/deserializing intervals, but I think that it may not be worth it - it sounds error-prone.

Reply:
@psarna the challenge is the interval consumers - most are not prepared for empty intervals.

Reply:
I have it all coded, but I found some very weird semantics of a nonwrapping_range (a.k.a. nowrapping_interval), and it's in the tests... In nonwrapping_range_test, we have:
```
BOOST_REQUIRE_EQUAL(r({5}, {1}).subtract(r({6}, {}), cmp), vec({r({}, {1}), r({5}, {{6, false}})}));
```
and my code fails with:
```
fatal error: in "test_range_subtract": critical check r({5}, {1}).subtract(r({6}, {}), cmp) == vec({r({}, {1}), r({5}, {{6, false}})}) has failed [{} != {(-inf, 1], [5, 6)}]
```
What the hell? How is a **non**wrapping range [5,1], which is empty, produce something nonempty when another range is subtracted from it? I also see that `partition_range` is a nonwrapping range, so perhaps this weird semantics is relied upon somewhere in the code.

Reply:
nonwrapping_range [5,1] is illegal under the current rules, not empty. So the test is illegal. It's not that the result is incorrect, the expression is undefined.

btw, please move the discussion out of this pull request. Github (or my browser) won't auto-scroll to the message when there is too much discussion, so I must scroll myself to find it.